infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,85 +101,127 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheConflictKeys.get(conflictKey);
     }
 
     /**
-     * Invalidate all records up to a trim mark.
-     *
-     * @param trimMark trim mark
+     * The first address in the priority queue.
      */
-    public void invalidateUpTo(long trimMark) {
-        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
-
-        AtomicLong entries = new AtomicLong();
+    public long firstAddress() {
+        if (cacheEntries.isEmpty()) {
+            return Address.NOT_FOUND;
+        }
+        return cacheEntries.peek().txVersion;
+    }
 
-        conflictCache.asMap().forEach((key, txVersion) -> {
-            if (txVersion == null || txVersion >= trimMark) {
-                return;
-            }
+    /**
+     * Invalidate the records with the minAddress. It could be one or multiple records
+     * @return the number of entries has been invalidated and removed from the cache.
+     */
+    private int invalidateFirst() {
+        ConflictTxStream firstEntry = cacheEntries.peek();
+        if (firstEntry == null) {
+            return 0;
+        }
 
-            conflictCache.invalidate(key);
-            entries.incrementAndGet();
-        });
+        int i = 0;
+        while (!cacheEntries.isEmpty() && cacheEntries.peek().txVersion == firstEntry.txVersion) {
+            ConflictTxStream entry = cacheEntries.poll();
+            cacheConflictKeys.remove(entry);
+            cacheEntriesBytes -= entry.size();
+            i++;
+        }
 
-        log.info("Invalidated entries: {}", entries.get());
+        log.trace("Evict {} entries", i);
+        maxConflictWildcard = Math.max(maxConflictWildcard, firstEntry.txVersion);
+        return i;
     }
 
     /**
-     * The cache size
+     * Invalidate all records up to a trim mark (not included).
      *
-     * @return cache size
+     * @param trimMark trim mark
      */
-    public long size() {
-        return conflictCache.estimatedSize();
+    public void invalidateUpTo(long trimMark) {
+        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
+        int entries = 0;
+        int pqEntries = 0;
+        while(Address.isAddress(firstAddress()) && firstAddress() < trimMark) {
+            pqEntries +=invalidateFirst();
```

Comment:
NIT: Add space after `while` and after `+=`

Reply:
Done.